Support stale samples for sparse histograms

codesome opened the pull request:
NOTE: this is pointing to sparsehistogram branch

This PR adds support for handling StaleNaN for sparse histograms. We accept stale samples via `Append()` too, so we do not require any special handling in scraping code.

cc @beorn7

beorn7 commented on tsdb/head_append.go:
```diff
@@ -605,7 +610,12 @@ func (s *memSeries) appendHistogram(t int64, sh histogram.SparseHistogram, appen
 
-	if !chunkCreated {
+	if !chunkCreated && value.IsStaleNaN(sh.Sum) {
+		// Deliberately empty other fields so that any future histogram will result in a new chunk.
```
Doesn't that mean that appending the stale marker itself will go into a new chunk?
Which would be wasteful, because at any regular end of a series, we would append a whole chunk with only the stale marker as the only content.

I think we either have to construct a histogram.SparseHistogram that is not seen as a counter reset (and ideally creates double-deltas of 0 to take the minimum amount of space), or we have to special-case the code within app.Appendable and app.AppendHistogram to deal with the stale-marker case appropriately.

codesome replied:
We could just use the last histogram with Sum set to StaleNaN then. And what special case do you have in mind? I did not exactly understand it.

beorn7 replied:
We could use the last histogram, with just the Sum changed, but that is unlikely to result in double-deltas of zero.

Because it's actually hard to create a histogram that will create double-deltas of zero, my idea was to special-case `app.Appendable` to always return "yup, it's appendable without problem" in case Sum is `StaleNaN`, and special-case `app.AppendHistogram` to just encode the Sum properly and set everything else to zero in case Sum is `StaleNaN`.

codesome replied:
But I believe if we get a new sample that is not stale, we do have to cut a new chunk at that time

beorn7 replied:
Yes, I think we should do that.

The idea is essentially to end the chunk with the stale marker.

But I think with the code as is in this PR, we would essentially create an own chunk thus for the stale marker.

beorn7 replied:
So perhaps we should have the following shortcut behavior in `app.Appendable` and `app.AppendHistogram`:

Method | Behavior if current sample is stale marker | Behavior if previous sample is stale marker
---|---|---
`app.Appendable` | Return `true` | return `false` (unless current sample is stale, too)
`app.AppendHistogram` | Append stale marker as `sum` and all other double-deltas and XOR's as zero | Append sample to a new chunk

codesome replied:
Yeah that is what I was trying to do, but facing some issues iterating it back. Will try it again tomorrow.

codesome commented:
@beorn7 I should be not cutting a new chunk for every stale sample now, ready for another review

beorn7 commented on tsdb/chunkenc/histo.go:
```diff
@@ -248,6 +249,15 @@ func (a *HistoAppender) Append(int64, float64) {}
 // * any buckets disappeared
 // * there was a counter reset in the count of observations or in any bucket, including the zero bucket
```
Perhaps the special behavior for the StaleNaN should be described in the doc comment here.

beorn7 reviewed (changes requested):
You wrote:
> We accept stale samples via Append() too, so we do not require any special handling in scraping code.

Do we, actually? See lines 237 to 241 in tsdb/chunkenc/histo.go. Don't we have to implement that method so that it calls the right other methods? Or perhaps we could move all the special casing over there? (Real question, I'm not sure…)

codesome commented:
> Do we, actually? See lines 237 to 241 in tsdb/chunkenc/histo.go. Don't we have to implement that method so that it calls the right other methods? Or perhaps we could move all the special casing over there? (Real question, I'm not sure…)

I see I have confused you a bit. We have 2 things, the headAppender and the HistoAppender for the Head block and the HistoChunk respectively, both having Append() and AppendHistogram(). What I am talking about is the headAppender, its Append() will accept StaleNaN and call headAppender.AppendHistogram, which will reduce code changes and special handling required in scraping. The behaviour of HistoAppender's Append() won't change, it will still a no-op.

beorn7 commented on tsdb/chunkenc/histo.go:
```diff
@@ -247,7 +248,17 @@ func (a *HistoAppender) Append(int64, float64) {}
 // * the zerobucket threshold has changed
 // * any buckets disappeared
 // * there was a counter reset in the count of observations or in any bucket, including the zero bucket
+// * the last sample in the chunk was stale while the current sample is not stale
```
Sorry for commenting yet again, but I think we should add that this method will always return `true` if the current sample is stale.

beorn7 reviewed (changes requested):
Thanks for the explanation. The doc comment in lines 237 to 241 in tsdb/chunkenc/histo.go should be updated to reflect that it should never be called. Or in fact, if it actually should never be called, let's panic if it is?

beorn7 commented:
Note: still looking into this, trying to understand the code in general (not this PR) better, so that I can hopefully give a few more useful comments.

beorn7 commented on tsdb/chunkenc/histo.go:
```diff
@@ -377,11 +394,14 @@ func (a *HistoAppender) AppendHistogram(t int64, h histogram.SparseHistogram) {
 		zcntDelta = int64(h.ZeroCount) - int64(a.zcnt)
 
+		if value.IsStaleNaN(h.Sum) {
+			tDelta, cntDelta, zcntDelta = 0, 0, 0
```
I think we still want to store the timestamp normally for a stale marker. If that's the case, `tDelta` should be treated normally and not set to 0 here.

codesome replied:
Good catch! But I am confused why are the tests not failing then. Let me find out...

codesome replied:
I get it now - the dod is 0 since timestamp are increasing at the same rate, hence not captured by tests.

beorn7 commented on tsdb/chunkenc/histo.go:
```diff
@@ -407,6 +428,10 @@ func (a *HistoAppender) AppendHistogram(t int64, h histogram.SparseHistogram) {
 		zcntDod := zcntDelta - a.zcntDelta
 
+		if value.IsStaleNaN(h.Sum) {
+			tDod, cntDod, zcntDod = 0, 0, 0
```
Similar to above, I believe `tDod` should stay as it is.

beorn7 commented on tsdb/head_append.go:
```diff
@@ -605,6 +610,11 @@ func (s *memSeries) appendHistogram(t int64, sh histogram.SparseHistogram, appen
 		return sampleInOrder, chunkCreated
 	}
 
+	if value.IsStaleNaN(sh.Sum) {
+		// Deliberately empty other fields so that any future histogram will result in a new chunk.
+		sh = histogram.SparseHistogram{Sum: sh.Sum}
+	}
```
Do we really need this? It seems the only way of receiving a stale marker here is via line 270 above, and that's already an empty histogram (with the exception of Sum).

codesome replied:
I don't want to rule out the possibility of getting it via headAppender.AppendHistogram() directly and cause regression later. That said, I think it is enough to take care of this in the histoChunk.AppendHistogram().

beorn7 commented on tsdb/chunkenc/histo.go:
```diff
@@ -350,6 +361,12 @@ func (a *HistoAppender) AppendHistogram(t int64, h histogram.SparseHistogram) {
 	var tDelta, cntDelta, zcntDelta int64
 	num := binary.BigEndian.Uint16(a.b.bytes())
 
+	if value.IsStaleNaN(h.Sum) {
+		// Emptying out other fields to write empty meta in case of
+		// first histogram in the chunk.
+		h = histogram.SparseHistogram{Sum: h.Sum}
+	}
```
I think the only code path that gives us a stale marker here is already ensuring that the histogram is otherwise empty. So we don't need this, do we?

codesome replied:
I plan to remove the cleanup in appendHistogram() in head.go and have it here since this chunk need not be used just by that code path. The case of adding stale via Append() is scraping specific.

beorn7 reviewed (changes requested):
OK, hopefully these comments are a bit smarter…

codesome commented:
@beorn7 I should have addressed your comments. I am still having the `h = histogram.SparseHistogram{Sum: h.Sum}` cleanup in the histo chunk and removed in other places. The natural cleanup that happens in headAppender.Append() I would take that as a scrape specific use case and would not rule out direct addition of stale samples via headAppender.AppendHistogram(), hence the cleanup in the chunk AppendHistogram().

beorn7 reviewed (approved):
OK, let's finally ship it! (o: 

I mean, we are not really shipping it yet. Thanks for your patience.
